### PR TITLE
FSDP2 Memory Tracker

### PR DESCRIPTION
**Pull request by sanketpurandare**
* __->__ #125323
### Why do we need the FSDP Memory Tracker?

**Tuning Decisions**

1. What is the expected peak memory with current configuration?
2. If I change my FSDP wrapping, how much effect will it have on peak memory?
3. What is the best batch size to use?
4. What is the maximum sequence length that one can run with current configuration?
5. How does increasing/decreasing the “DP” world size affect peak memory?
6. How much memory do I save if I move the optimizer to the CPU?
7. Which activation checkpointing policy should I use?
8. If I have various SAC policies, How do they compare against each other?
9. What happens if I apply different SAC policies to different FSDP units?
10. If I make my gradient reduction in fp32, what effect will it have on memory?
11. If I want to use a custom mixed precision policy, how will it affect the peak memory?
12. When does it make sense to use HSDP?
13. Can I reshard to a smaller mesh without increasing peak memory substantially?
14. Can safely disable post forward reshard without causing an OOM?

**Debugging**

1. Which module contributes most to activation memory?
2. Which FSDP unit is holding a lot of unsharded memory?
3. AC is not releasing memory?

The FSDP2 Memory Tracker addresses all of the above. It is based on:
 *  #124688 
 *  #128508  

Example and Output:

```
if __name__== "__main__":
    from contextlib import nullcontext
    from functools import partial
    import torch
    from torch.distributed._composable import checkpoint
    from torch.distributed._composable.fsdp import (
        CPUOffloadPolicy,
        fully_shard,
        MixedPrecisionPolicy,
    )
    from torch.distributed._tensor import DeviceMesh
    from torch.distributed._tools.fsdp2_mem_tracker import FSDPMemTracker
    from torch._subclasses.fake_tensor import FakeTensorMode
    from torch.testing._internal.distributed._tensor.common_dtensor import (
    ModelArgs,
    Transformer,
    TransformerBlock,
    )
    from torch.testing._internal.distributed.fake_pg import FakeStore
    dev = torch.device("cuda:0")
    torch.cuda.set_device(dev)
    world_size = 4
    store = FakeStore()
    torch.distributed.init_process_group(
        "fake", rank=0, world_size=world_size, store=store
    )
    mesh = DeviceMesh("cuda", torch.arange(0, world_size))
    torch.cuda.empty_cache()
    torch.manual_seed(42)
    use_fake_mode = False
    with FakeTensorMode() if use_fake_mode else nullcontext():
        vocab_size = 8192
        bsz, seq_len = 32, 1024
        with torch.device(dev):
            model_args = ModelArgs(
                n_layers=2,
                n_heads=16,
                vocab_size=vocab_size,
                max_seq_len=seq_len,
                dropout_p=0.1,
            )
            model = Transformer(model_args)
        foreach = True
        mp_policy = MixedPrecisionPolicy(param_dtype=torch.bfloat16, reduce_dtype=torch.float32)
        offload_policy = CPUOffloadPolicy(pin_memory=not use_fake_mode)
        reshard_after_forward = True
        fsdp_config = {

        }
        fully_shard_fn = partial(
            fully_shard,
            mesh=mesh,
            reshard_after_forward=reshard_after_forward,
            offload_policy=offload_policy,
            mp_policy=mp_policy,
        )
        for module in model.modules():
            if isinstance(module, TransformerBlock):
                checkpoint(module, preserve_rng_state=not use_fake_mode)
                fully_shard_fn(module)
        fully_shard_fn(model)
        optim = torch.optim.Adam(model.parameters(), lr=1e-2, foreach=foreach)

        torch.manual_seed(42)
        inp = torch.randint(0, vocab_size, (bsz, seq_len), device=dev)
        torch.cuda.reset_accumulated_memory_stats()
        torch.cuda.reset_peak_memory_stats()
        fmt = FSDPMemTracker(model, optim)
        fmt.track_inputs((inp,))
        with fmt:
            for iter_idx in range(2):
                loss = model(inp).sum()
                loss.backward()
                optim.step()
                optim.zero_grad()
                if iter_idx == 0:
                    fmt.reset_mod_stats()
    mem_stats = torch.cuda.memory_stats()
    tracker_peak = fmt.get_tracker_snapshot("peak")[dev]["Total"]
    cuda_peak_active = mem_stats["active_bytes.all.peak"]
    fmt.display_modulewise_snapshots(depth=4, units="MiB", tabulate=True)
    fmt.display_snapshot("peak", units="MiB", tabulate=True)
    print(
        f"peak active: {cuda_peak_active / (1024**3)} GiB | "
        f"Tracker Max: {tracker_peak / (1024 ** 3)} GiB"
    )
    if not use_fake_mode:
        print(f"Accuracy: {tracker_peak/cuda_peak_active}")

    try:
        torch.distributed.destroy_process_group()
    except Exception as e:
        print(e)
```

Screenshot 2024-06-21 at 5 16 49 PM

cc: @gnadathur @awgu @wanchaol @weifengpy 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

### CODE REVIEW

**Comment by awgu** [comment]
High-level comment: It might be worthwhile to document which parts are depending on FSDP2 internals, and we may be able to see how to expose things more robustly.

File: torch/distributed/_tools/fsdp2_memory_tracker.py [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+
+class _FSDPParamGroupSavedMethods(NamedTuple):
+    pre_forward: Callable
+    post_forward: Callable
+    pre_backward: Callable
+    post_backward: Callable
```
Comment (awgu) [inline torch/distributed/_tools/fsdp2_memory_tracker.py:72]: These methods should be pretty stable, so I am good with patching/introspecting these.

File: torch/distributed/_tools/fsdp2_memory_tracker.py [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+        orig_fsdp_param_group_pre_forward: Callable,
+        fsdp_param_group: FSDPParamGroup,
+        name: str,
+    ) -> Callable:
+        @wraps(orig_fsdp_param_group_pre_forward)
+        def inner(*args, **kwargs):
+            self._enter_module(name, "Before Pre-Forward")
+            args, kwargs = orig_fsdp_param_group_pre_forward(*args, **kwargs)
+            for fsdp_param in fsdp_param_group.fsdp_params:
+                assert isinstance(fsdp_param._unsharded_param, nn.Parameter), (
```
Comment (awgu) [inline torch/distributed/_tools/fsdp2_memory_tracker.py:243]: Only nit from our discussion was to try to use the public `fsdp_param.unsharded_param` here instead of `_unsharded_param` since it will include a check that the parameter is in the unsharded state.

File: test/distributed/_tools/test_fsdp2_mem_tracker.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    ModelArgs,
+    Transformer,
+    TransformerBlock,
+)
+
+c10d_ops = torch.ops.c10d
+funcol = torch.ops.c10d_functional
+
+
```
Comment (sanketpurandare) [inline test/distributed/_tools/test_fsdp2_mem_tracker.py:33]: Test sizes are purposefully kept a bit large to stress test the tracker, but I have made sure all of them have peak memory below 3GB and run fast enough.

File: test/distributed/_tools/test_fsdp2_mem_tracker.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+class TestTrackerFullyShard1DTrainingCore(FSDPTest):
+    @property
+    def world_size(self) -> int:
```
Comment (awgu) [inline test/distributed/_tools/test_fsdp2_mem_tracker.py:36]: I think if you do not add this to run in a multi-GPU test, then it will never have 4 GPUs (only 2).

See
https://github.com/pytorch/pytorch/blob/9d1b65b569a304032776375c24aa6babbf576c30/.ci/pytorch/test.sh#L325-L326
for where we sneak in some tests to run with Inductor tests that have 4 GPUs.

File: test/distributed/_tools/test_fsdp2_mem_tracker.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    @skip_if_lt_x_gpu(2)
+    def test_tracker_multi_group_eager(self):
+        """
+        Tests tracker accuracy when using multiple parameter groups for
+        communication (for communication and computation overlap plus memory
+        reduction) and different mixed precision policies.
+        """
+        self.run_subtests(
+            {
+                "reshard_after_forward": [2],
```
Comment (awgu) [inline test/distributed/_tools/test_fsdp2_mem_tracker.py:48]: E.g., if you do `reshard_after_forward=2` but only have world size 2, then it would get clamped to `True`.

File: test/distributed/_tools/test_fsdp2_mem_tracker.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                    CPUOffloadPolicy(pin_memory=False),
+                    OffloadPolicy(),
+                ],
+                "mp_policy": [
+                    MixedPrecisionPolicy(
+                        param_dtype=torch.float16, reduce_dtype=torch.float32
+                    ),
+                    MixedPrecisionPolicy(),
+                ],
+                "device_type": ["cuda"],
```
Comment (awgu) [inline test/distributed/_tools/test_fsdp2_mem_tracker.py:59]: nit: I think I included `device_type` before in case we cared about CPU training, but it turns out that we are pretty far from supporting that (a few different things we need to address). I guess you could exclude this arg and just directly use `"cuda"`.

File: test/distributed/_tools/test_fsdp2_mem_tracker.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+        mp_policy: MixedPrecisionPolicy,
+        device_type: str,
+    ):
+        assert device_type in ("cuda",), f"{device_type}"
+        dev = torch.device(torch.cuda.current_device())
+        torch.cuda.empty_cache()
+        torch.manual_seed(42)
+        lin_dim, bsz = 2048, 8192
+        model = nn.Sequential(*[MLP(lin_dim, torch.device("cpu")) for _ in range(4)])
+        mesh = init_device_mesh(device_type, (self.world_size,))
```
Comment (awgu) [inline test/distributed/_tools/test_fsdp2_mem_tracker.py:77]: yea it seems like we can just de-dup this and use CUDA directly since you are using it unconditionally for the `inp` anyway

File: test/distributed/_tools/test_fsdp2_mem_tracker.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        fmt = FSDPMemTracker(model, optim)
+        fmt.track_inputs((inp,))
+        with fmt:
+            for iter_idx in range(2):
+                loss = model(inp).sum()
+                loss.backward()
+                optim.step()
+                optim.zero_grad()
+                if iter_idx == 0:
+                    fmt.reset_mod_stats()
```
Comment (awgu) [inline test/distributed/_tools/test_fsdp2_mem_tracker.py:101]: Why do we need this reset?

File: torch/distributed/_tools/fsdp2_mem_tracker.py [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+                    snapshot
+                )
+                mod_stat.snapshots.setdefault(_FSDPModState.BEF_PRE_FW, []).append(
+                    deepcopy(snapshot)
+                )
+            elif not self._mod_tracker.is_bw:
+                parents = self._mod_tracker.parents - {mod_fqn}
+                if len(parents) == 1 and "Global" in parents:
+                    raise NotImplementedError(
+                        "FSDPMemTracker does not support memory tracking for multiple iterative calls."
```
Comment (awgu) [inline torch/distributed/_tools/fsdp2_mem_tracker.py:223]: I am slow -- what does multiple iterative calls mean?

File: torch/distributed/_tools/fsdp2_mem_tracker.py [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+
+    def _fsdp_param_group_pre_backward(
+        self,
+        fsdp_mod: FSDPModule,
+        orig_fsdp_param_group_pre_backward: Callable,
+    ) -> Callable:
+        # We capture memory snapshots before and after ``FSDPParamGroup.pre_backward`` to capture the pre-fetching
+        # and unsharding of params. We also initialize ``local_peak`` and ``PEAK_BW`` snapshot for the module.
+        @wraps(orig_fsdp_param_group_pre_backward)
+        def inner(*args: Any) -> None:
```
Comment (awgu) [inline torch/distributed/_tools/fsdp2_mem_tracker.py:290]: Should we just use `*args` and `**kwargs` to be safe? *(also for post-backward) E.g., we might later choose to call `FSDPParamGroup.post_backward` with a kwarg.

File: torch/distributed/_tools/fsdp2_mem_tracker.py [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+    ) -> Callable:
+        # We capture the memory snapshots before and after ``FSDPParamGroup.post_backward`` to track and attribute
+        # the `unsharded` grads before the post backward and then `sharded` grads and `reduce_scatter`  buffers
+        # after the post backward.
+        @wraps(orig_fsdp_param_group_post_backward)
+        def inner(*args: Any) -> None:
+            fsdp_state = fsdp_mod._get_fsdp_state()
+            if fsdp_param_group := fsdp_state._fsdp_param_group:
+                for fsdp_param in fsdp_param_group.fsdp_params:
+                    unsharded_grad = fsdp_param._unsharded_param.grad
```
Comment (awgu) [inline torch/distributed/_tools/fsdp2_mem_tracker.py:321]: I think this will need to get updated in the future for fp32 grad accumulation.

**Review by awgu — APPROVED** [review]
This seems okay to me! I think we just need to be careful of the situation "I changed something in FSDP2 and now it broke the memory tracker. How do we fix it?"

File: torch/distributed/_tools/fsdp2_mem_tracker.py (hunk 9 above)
Comment (sanketpurandare) [reply torch/distributed/_tools/fsdp2_mem_tracker.py:223]: I mean if someone calls the `train_step()` multiple times in a loop, then they need to use `reset_mod_stats()` after each iteration, because we cannot aggregate module memory stats across multiple iterations. Peak across multiple iterations will be maintained.

File: torch/distributed/_tools/fsdp2_mem_tracker.py (hunk 11 above)
Comment (sanketpurandare) [reply torch/distributed/_tools/fsdp2_mem_tracker.py:321]: Do you think using `register_hook` or `register_post_accumulate_grad_hook` is better?

File: test/distributed/_tools/test_fsdp2_mem_tracker.py (hunk 8 above)
Comment (sanketpurandare) [reply test/distributed/_tools/test_fsdp2_mem_tracker.py:101]: https://github.com/pytorch/pytorch/pull/125323#discussion_r1649477933

**Comment by sanketpurandare** [comment]
> This seems okay to me! I think we just need to be careful of the situation "I changed something in FSDP2 and now it broke the memory tracker. How do we fix it?"

Yeah we need to maintain it actively till FSDP2 stabilizes.

File: torch/distributed/_tools/fsdp2_mem_tracker.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/_tools/fsdp2_mem_tracker.py:223]: oh I see

File: torch/distributed/_tools/fsdp2_mem_tracker.py (hunk 11 above)
Comment (awgu) [reply torch/distributed/_tools/fsdp2_mem_tracker.py:321]: oh the issue for fp32 grad accumulation is that we may store the fp32 unsharded grad in a different attribute

I can sync with you on this later offline

File: test/distributed/_tools/test_fsdp2_mem_tracker.py (hunk 4 above)
Comment (sanketpurandare) [reply test/distributed/_tools/test_fsdp2_mem_tracker.py:36]: For now, I will skip adding anything to CI. Once people use the tracker and we harden it, then we can add more tests.

**Comment by huydhn** [comment]
@sanketpurandare I'm seeing the new test `test_tracker_multi_group_eager` failing on ROCm distributed job https://hud.pytorch.org/pytorch/pytorch/commit/287c68c5eca2e15bf73b84fe9e39755ae3f842ba#26578545778.  Could you help take a look?  The job is only run periodically, do its signal was missed on the PR

**Comment by sanketpurandare** [comment]
> @sanketpurandare I'm seeing the new test `test_tracker_multi_group_eager` failing on ROCm distributed job https://hud.pytorch.org/pytorch/pytorch/commit/287c68c5eca2e15bf73b84fe9e39755ae3f842ba#26578545778. Could you help take a look? The job is only run periodically, do its signal was missed on the PR

Yeah will push a fix

**Comment by huydhn** [comment]
Btw, I disable the test in https://github.com/pytorch/pytorch/issues/129390 to keep trunk sane.  In your fixed PR, please add "Fixed https://github.com/pytorch/pytorch/issues/129390" in yout PR description to run the test in your PR

**Comment by sanketpurandare** [comment]
> Btw, I disable the test in #129390 to keep trunk sane. In your fixed PR, please add "Fixed #129390" in yout PR description to run the test in your PR

Yeah it's because of Python 3.8 not supporting bitwise OR in typing
